doc/decisions/0_drafts/keyname.md

```diff
@@ -0,0 +1,136 @@
+
+- Because `KeySet` is ordered by name and stores `Key`, the order comparison between the name of two `Key`s must be fast.
+
+## Assumptions
+
+- In most cases the escaped name is used for developer convenience not, because of actual requirements.
+- The most common requirement for using the escaped name is UI: reading names from or displaying them in a user interface (e.g., `kdb` CLI)
+
+## Considered Alternatives
+
+### Only escaped name
+
+Because the escaped name is a simple zero-terminated string, it can be represented as a single `char *`.
+
+Storing the name as a single `char *` would be the most space efficient.
+But resizing would require counting the length every time.
+Therefore, for storage the better solution may be a `char *` and a `size_t`.
+
+However, in the API the name could always be a single `char *`, making for a very easy to use API.
+
+The biggest problem with this approach is that comparing two escaped names is not trivial.
+The comparison needs to account for namespaces, parts and escaping.
+
+Similarly, iterating over the individual parts of a name (and/or manipulating them) is non-trivial, because it requires logic to handle escape sequences.
+
+### Only unescaped name
+
+The unescaped name contains zero bytes.
+It therefore must be represented as a pointer and a size.
+
+This can make for less convenient API, but there are mitigation strategies using additional types.
+Using unescaped names in code can be inconvenient, especially regarding the namespace.
+Without a namespace a name could be written as e.g., `"foo\0bar"`.
+But with a namespace it would be something like `"\1\0\foo\0bar"` and developers would need to remember what namespace `\1` is.
+Using the `KEY_NS_*` constants like this is not easily possible.
+
+Both order and hierarchy comparisons are very simple in this case and can be implemented with little more logic than a few `memcmp`.
+Iterating over the individual parts is also trivial, since all parts are separated by zero bytes.
+
+### Only unescaped name, with separate namespace
+
+To solve the namespace issue of the above solution, the namespace could be separated.
+
+In the API it is an obvious solution.
+Instead of taking a single pointer and size, which receive values like `"\1\0foo\0bar"` and `10`, the API would take a namespace, a pointer, and a size, with values like `KEY_NS_CASCADING`, `"foo\0bar"` and `8`.
+
+Internally, this could be combined into one buffer, stored as a single pointer and size, again.
+However, depending on the API there can also be benefits to keeping the namespace as a separate field.
+
+Even with a separate namespace field, the benefits of "Only unescaped name" are retained.
+The memory consumption is near minimal (alignment padding can cause a difference).
+Comparisons are exactly the same, just with an additional namespace byte comparison beforehand.
+
+### Both escaped and unescaped name
+
+To combine the advantages of escaped and unescaped name, both could be used.
```

Comment:
This is the current approach, right?

Reply:
Yes

Reply:
I can add it, but I don't know, will anybody really care about that in a few months/years? When all this is done and implemented, nobody will care what the previous implementation was IMO.